test/suite/lib.rs

```diff
@@ -63,74 +63,100 @@ async fn assert_balance(
+    )
+    .await
+    .with_context(|| {
+        format!(
+            "failed to transfer {} tokens to {}",
+            transfer_amount,
+            p2_env.principal()
+        )
+    })?;
 
     assert_balance(
-        &ledger_env,
+        &p2_env,
         Account {
-            owner: receiver,
+            owner: p2_env.principal(),
             subaccount: None,
         },
-        amount,
+        balance_p2.clone() + Nat::from(transfer_amount),
     )
     .await?;
 
-    assert_balance(&ledger_env, Account{ owner: receiver, subaccount: Some([0; 32]) }, amount)
+    assert_balance(&ledger_env, Account{ owner: p2_env.principal(), subaccount: Some([0; 32]) }, balance_p2 + transfer_amount)
         .await
         .context("the ledger does not treat accounts with an empty subaccount as accounts with the default subaccount")?;
 
+    assert_balance(
+        &p1_env,
+        p1_env.principal(),
+        balance_p1 - Nat::from(transfer_amount) - transfer_fee(&p1_env).await?,
+    )
+    .await?;
+
     Ok(Outcome::Passed)
 }
 
 /// Checks whether the ledger supports token burns.
-/// Skips the checks if the ledger does not have a minting account.
-pub async fn test_burn(ledger_env: impl LedgerEnv + LedgerEnv) -> TestResult {
-    let minting_account = match minting_account(&ledger_env).await? {
-        Some(account) => account,
-        None => {
-            return Ok(Outcome::Skipped {
-                reason: "the ledger does not support burn transactions".to_string(),
-            });
-        }
-    };
-
-    assert_balance(&ledger_env, minting_account.clone(), 0)
-        .await
-        .context("minting account cannot hold any funds")?;
-
-    let burn_amount = Nat::from(10_000) + transfer_fee(&ledger_env).await?;
-    let tmp_account_env = ledger_env.fork();
-    let tmp_account = tmp_account_env.principal();
-
-    transfer(
-        &ledger_env,
-        Transfer::amount_to(burn_amount.clone(), tmp_account),
-    )
-    .await?
-    .context("failed to transfer funds to a temporary account")?;
-
-    assert_balance(&ledger_env, tmp_account, burn_amount.clone()).await?;
-
-    let _tx = transfer(
-        &tmp_account_env,
-        Transfer::amount_to(burn_amount, minting_account.clone()).fee(0),
-    )
-    .await
-    .context("failed to burn tokens");
-
-    assert_balance(&ledger_env, tmp_account, 0).await?;
-    assert_balance(&ledger_env, minting_account, 0).await?;
-
+/// Expects the given account to have a balance of at least 2*Transfer_Fee
+pub async fn test_burn(ledger_env: impl LedgerEnv + LedgerTransaction + Clone) -> TestResult {
+    let burn_amount = Nat::from(10_000);
+    let envs = setup_test_accounts(&ledger_env, burn_amount.clone(), 1).await?;
+    let p1_env = envs[0].clone();
+
+    p1_env
+        .burn(burn_amount.clone())
+        .await?
+        .context("failed to burn amount")?;
+
+    assert_balance(&p1_env, p1_env.principal(), 0).await?;
```

Comment:
Shouldn't this be `previous balance - burn_amount` instead of `0`? You don't know if the account had tokens in it. It's unlikely but not impossible. We can use a generic invariant that works regardless of the balance of the account before the `setup_test_accounts`.

Reply:
When setting up the testing account, I am checking whether the newly generated account has a balance of 0, so before this assert, I can be sure that the account had 0 tokens at the beginning of the test. I then transfer 10_000 tokens to it and burn them again. That is why I check for a balance 0. Of course, it is theoretically possible that the random number generator generates an account which already exists and has non zero balance or that it is an active account and transfers are being made while I a running the test, but I would categorize this as a very low possibility as the numeric space of random numbers for u64 is pretty large.